src/builder/modules/Generator_Builder.res

```diff
@@ -437,6 +437,66 @@ let fillBackgroundColor = (model: Model.t, fillStyle: string) => {
   }
 }
 
+let getOffset = ((x1, y1): position, (x2, y2): position) => {
+  let x1 = Belt.Int.toFloat(x1)
+  let y1 = Belt.Int.toFloat(y1)
+  let x2 = Belt.Int.toFloat(x2)
+  let y2 = Belt.Int.toFloat(y2)
+
+  let w = x2 -. x1
+  let h = y2 -. y1
+
+  /* When a line is drawn and its start and end coordinates are integer values, 
+  the resulting line is drawn in between to rows of pixels, resulting in a line
+  that is two pixels wide and half transparent. To fix this, the line's start
+  and end positions need to be offset 0.5 pixels in the direction normal to the
+  line's slope. The following code gets the angle of the line, and creates the
+  offset needed in each direction to place the line in the correct spot. This
+  results in a fully opaque line with the correct width if the line is vertical
+  or horizontal, but antialiasing may still affect lines at other angles.
```

Comment:
Great comment.

Do you have a reference to a page that explains the maths? Or maybe you worked this out this on your own? If you have a reference, can you add that here as well?

It's a great solution.